Conflicts UI additions

mcginty opened the pull request.

moxie0 commented on res/layout/conversation_item_sent.xml:
```diff
@@ -62,7 +66,8 @@
                 android:textColorLink="?conversation_sent_text_primary_color"
-                android:textSize="16sp" />
+                android:textSize="16sp"
+                tools:text="Lorem ipsum mango dolor coconut papaya" />
```
?

moxie0 commented on res/layout/message_details_header.xml:
```diff
@@ -0,0 +1,76 @@
+                  android:layout_width="wrap_content"
+                  android:layout_height="wrap_content"
+                  android:layout_marginLeft="@dimen/message_details_table_row_pad"
+                  tools:text="Jan 18, 9:29AM" />
```
?

moxie0 commented on res/layout/message_details_header.xml:
```diff
@@ -0,0 +1,76 @@
+                  android:layout_width="wrap_content"
+                  android:layout_height="wrap_content"
+                  android:layout_marginLeft="@dimen/message_details_table_row_pad"
+                  tools:text="Jan 18, 9:31AM" />
```
?

moxie0 commented on res/values/strings.xml:
```diff
@@ -72,7 +72,7 @@
     <string name="ConversationItem_message_size_d_kb">Message size: %d KB</string>
     <string name="ConversationItem_expires_s">Expires: %s</string>
-    <string name="ConversationItem_error_sending_message">Error sending message</string>
+    <string name="ConversationItem_error_sending_message">Not delivered</string>
```
string name needs to change too then

moxie0 commented on src/org/thoughtcrime/securesms/ConversationItem.java:
```diff
@@ -229,7 +221,7 @@ private void setConversationBackgroundDrawables(MessageRecord messageRecord) {
+        if ((messageRecord.isPending() || messageRecord.isFailed()) && pushDestination && !messageRecord.isForcedSms()) {
           background         = SENT_PUSH_PENDING;
           triangleBackground = SENT_PUSH_PENDING_TRIANGLE;
         } else if (messageRecord.isPending() || messageRecord.isPendingSmsFallback()) {
```
You don't want to add failed to this one too?

mcginty replied:
now that you mention it...

moxie0 commented on src/org/thoughtcrime/securesms/MessageDetailsActivity.java:
```diff
@@ -0,0 +1,182 @@
+    pushDestination = getIntent().getBooleanExtra(PUSH_EXTRA, false);
+  }
+
+  private void initializeDates() {
```
It's called initializeDates, but initializes both transport and date values.

moxie0 commented on src/org/thoughtcrime/securesms/MessageDetailsActivity.java:
```diff
@@ -0,0 +1,182 @@
+
+  private MasterSecret     masterSecret;
+  private MessageRecord    messageRecord;
+  private ConversationItem item;
```
i'd prefer conversationItem or anything more descriptive

moxie0 commented on src/org/thoughtcrime/securesms/MessageDetailsActivity.java:
```diff
@@ -0,0 +1,182 @@
+  private TextView         transport;
+  private TextView         toFrom;
+  private ListView         recipientsList;
+  private LayoutInflater   li;
```
we are not heathens

moxie0 commented on src/org/thoughtcrime/securesms/MessageDetailsRecipient.java:
```diff
@@ -0,0 +1,179 @@
+    this.resendButton      = (Button)    findViewById(R.id.resend_button);
+  }
+
+  public void set(final MasterSecret masterSecret, final MessageRecord record, final Recipients recipients, final int position) {
```
this method is way too long

mcginty replied:
refactored

moxie0 commented on src/org/thoughtcrime/securesms/MessageDetailsActivity.java:
```diff
@@ -0,0 +1,182 @@
+  }
+
+  @Override
+  public Loader<Pair<MessageRecord,Recipients>> onCreateLoader(int id, Bundle args) {
```
Is there a compelling reason to use a loader and all the infrastructure around that, rather than just running an asynctask?

mcginty replied:
As far as I know, Loaders take care of some of the special cases for the Activity being destroyed on orientation changes, etc, making them useful in avoiding some edge case bugs.

moxie0 replied:
We don't do it anywhere else for non-lists.  Seems like a lot of boilerplate for something that could be a 5-line async task.

moxie0 commented on src/org/thoughtcrime/securesms/database/Database.java:
```diff
@@ -25,9 +25,9 @@
-  private static final String CONVERSATION_URI      = "content://textsecure/thread/";
-  private static final String CONVERSATION_LIST_URI = "content://textsecure/conversation-list";
+  protected static final String ID_WHERE              = "_id = ?";
+  public    static final String CONVERSATION_URI      = "content://textsecure/thread/";
```
why  make this public?

mcginty replied:
The MessageDetailsActivity listens on changes to the conversation so it can change the message details accordingly.

moxie0 replied:
I don't think we should expose that.  If we need to refresh it should be holding onto a cursor.

moxie0 commented on src/org/thoughtcrime/securesms/MessageDetailsActivity.java:
```diff
@@ -0,0 +1,173 @@
+    pushDestination = getIntent().getBooleanExtra(PUSH_EXTRA, false);
+  }
+
+  private void initializeTimeAndTransport() {
```
hah, i meant this should be two separate methods

moxie0 replied:
Also, it's a little weird to have initialize methods that aren't called in onCreate

moxie0 commented on src/org/thoughtcrime/securesms/MessageDetailsRecipient.java:
```diff
@@ -0,0 +1,198 @@
+  }
+
+  private boolean populateConflicts(final MasterSecret masterSecret, final MessageRecord record) {
+    boolean applicableFailures = false;
```
cut and paste huh.

these types of break variables always feel fishy to me too.

moxie0 commented on src/org/thoughtcrime/securesms/ConversationFragment.java:
```diff
@@ -101,11 +99,11 @@ private void initializeResources() {
   }
 
   private void initializeListAdapter() {
+    pushDestination = DirectoryHelper.isPushDestination(getActivity(), this.recipients);
```
The fact that we've never had a state variable for this before makes me uneasy to introduce it now.  Seems like the kind of thing that will fall out of sync.

mcginty commented:
addressed comments, everything's in an asynctask sans loader boilerplate

moxie0 commented:
I might not be reading the diff right, but my recollection of our in-person conversation about this was that you would use a standard cursoradapter to map the message status, so that you didn't have to expose any of the ContentObserver stuff to a view.  That way you'd use an AsyncTask to resolve the Recipient information the adapter returned?

mcginty commented:
Yeah, you know what? I did that and it wasn't working, but realized that it will now after I added the content URIs to the cursors in `getMessage`. Fixed.

moxie0 commented on src/org/thoughtcrime/securesms/MessageDetailsActivity.java:
```diff
@@ -0,0 +1,184 @@
+    }
+  }
+
+  private void inflateHeaderIfAbsent() {
```
Is there a reason not to do this at create time?

mcginty replied:
Ah, I think initially it depended on the results of the AsyncTask. No longer. Moved with the rest of the resource initialization code.

moxie0 commented on src/org/thoughtcrime/securesms/MessageDetailsRecipientAdapter.java:
```diff
@@ -0,0 +1,59 @@
+  private Recipients    recipients;
+
+  public MessageDetailsRecipientAdapter(Context context, MasterSecret masterSecret, MessageRecord record, Recipients recipients) {
+    super();
```
unnecessary

moxie0 commented on src/org/thoughtcrime/securesms/MessageDetailsRecipient.java:
```diff
@@ -0,0 +1,205 @@
+      @Override
+      public void run() {
+        fromView.setText(formatFrom(recipient));
+        setContactPhoto(recipient);
```
feels a little code duplicationey

moxie0 commented on src/org/thoughtcrime/securesms/MessageDetailsRecipient.java:
```diff
@@ -0,0 +1,205 @@
+ *
+ * @author Jake McGinty
+ */
+public class MessageDetailsRecipient extends RelativeLayout
```
MessageDetailsRecipientView?

mcginty replied:
Sure. We may want to rename Conversation(List)Item to be Conversation(List)ItemView too to maintain consistency.

moxie0 replied:
I'm also fine with Item as a suffix

moxie0 commented on src/org/thoughtcrime/securesms/MessageRecipientAsyncTask.java:
```diff
@@ -0,0 +1,120 @@
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.thoughtcrime.securesms;
```
Is this the right package?

mcginty replied:
moved to util, that or I can move as a subclass of MessageDetailsActivity, but I think I personally tend against classes with large-ish nontrivial subclasses just for readability.

moxie0 commented on src/org/thoughtcrime/securesms/MessageRecipientAsyncTask.java:
```diff
@@ -0,0 +1,120 @@
+import java.lang.ref.WeakReference;
+import java.util.LinkedList;
+
+public abstract class MessageRecipientAsyncTask extends AsyncTask<Void,Void,Result> {
```
This is written as an abstract class, but is there ever a situation where this would be used again?

moxie0 replied:
I like the idea of an AsyncTask that resolves recipient information, but is there a reason to put the cursor parsingn interaction in here?

What if this had a constructor that accepted a Context and  MasterSecret, an execute() method that accepted a MessageRecord, and returned a Recipients object?

mcginty replied:
I marked it as abstract because it's incomplete as there's no `onPostExecute()`, so it's useless unless a superclass implements that.

Yeah that's cleaner to do MessageRecord resolving outside the AsyncTask so it can just return Recipients. Fixed.

It always seemed weird to me to have an AsyncTask where execute() takes a list of arguments and then have code that forces it to be singular, but changed anyway.

moxie0 replied:
Doesn't look like the new code made it up?

mcginty commented:
I was refactoring the various list items per your duplicatey comment. There's a decent amount of common code between list items that refer to recipients, so I created a more DRY setup.
